File: src/main/java/dp/max/MaxSubArraySum.java
package dp.max;

/**
 * @author ShaneTang
 * @create 2021-02-21 17:33
 */
public class MaxSubArraySum {

    public static void main(String[] args) {
        MaxSubArraySum obj = new MaxSubArraySum();
        int res = obj.solutionGreedy(new int[]{1, -2, 3, 10, -4, 7, 2, -5});
        System.out.println("res = " + res);
    }

    public int dpTable(int[] nums) {
        int n = nums.length;
        int res = nums[0];
        int[] dp = new int[n];
        dp[0] = nums[0];
        for (int i = 1; i < n; i++) {
            dp[i] = Math.max(nums[i] + dp[i - 1], nums[i]);
            res = Math.max(res, dp[i]);
        }
        return res;
    }

    public int solutionGreedy(int[] nums) {
        return new greedy.MaxSubArraySum().solutionGreedy(nums);
    }

    public int solutionDpOpt(int[] nums) {
        return new greedy.MaxSubArraySum().solutionDpOpt(nums);
    }
}

File: src/main/java/dp/max/max_subarray_sum.go
package main

func main() {
	println(dpTable([]int{-2, 1, -3, 4, -1, 2, 1, -5, 4}))
}

func maxSubArray(nums []int) int {
	res := nums[0]
	curSum := 0
	for _, val := range nums {
		curSum += val
		res = max(res, curSum)
		if curSum < 0 {
			curSum = 0
		}
	}
	return res
}

func dpTable(nums []int) int {
	//当 dp[i - 1] > 0 dp[i−1]>0 时：执行 dp[i] = dp[i-1] + nums[i] ；
	//当 dp[i - 1] < 0 dp[i−1]≤0 时：执行 dp[i] = nums[i] ；
	res := nums[0]
	dp := make([]int, len(nums))
	dp[0] = nums[0]
	for i := 1; i < len(dp); i++ {
		dp[i] = max(dp[i-1]+nums[i], nums[i])
		res = max(res, dp[i])
	}
	return res

}

func maxOfSlice(slice []int) int {
	max := slice[0]
	for _, val := range slice {
		if val > max {
			max = val
		}
	}
	return max
}
